File: epi_judge_python/sorted_lists_merge.py
from typing import Optional

from list_node import ListNode
from test_framework import generic_test


def merge_two_sorted_lists(
    L1: Optional[ListNode], L2: Optional[ListNode]
) -> Optional[ListNode]:
    if not L1 and not L2:
        return None
    if not L1:
        return L2
    elif not L2:
        return L1
    else:
        if L1.data < L2.data:
            head = sm = L1
            lr = L2
        else:
            head = sm = L2
            lr = L1
    while sm and lr:
        while sm.next and sm.data == sm.next.data:
            sm = sm.next
        while sm.next and sm.next.data <= lr.data:
            sm = sm.next
        tmp = sm.next
        sm.next, lr = lr, tmp
        sm = sm.next
    return head


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "sorted_lists_merge.py", "sorted_lists_merge.tsv", merge_two_sorted_lists
        )
    )

File: epi_judge_cpp/sorted_lists_merge.cc
#include "list_node.h"
#include "test_framework/generic_test.h"
#include <memory>
shared_ptr<ListNode<int>> MergeTwoSortedLists(shared_ptr<ListNode<int>> L1,
                                              shared_ptr<ListNode<int>> L2) {
  // auto head = L1->data < L2->data
  //                 ? std::make_shared<ListNode<int>>(L1->data, nullptr)
  //                 : std::make_shared<ListNode<int>>(L2->data, nullptr);
  auto head = std::make_shared<ListNode<int>>();
  auto tail = head;
  auto append_node = [&tail](shared_ptr<ListNode<int>> &node) {
    // tail->next = std::make_shared<ListNode<int>>(node->data);
    tail->next = node;
    node = node->next;
    tail = tail->next;
  };

  while (L1 && L2) {
    append_node(L1->data < L2->data ? L1 : L2);
  }
  tail->next = L1 ? L1 : L2;
  return head->next;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L1", "L2"};
  return GenericTestMain(args, "sorted_lists_merge.cc",
                         "sorted_lists_merge.tsv", &MergeTwoSortedLists,
                         DefaultComparator{}, param_names);
}
